plugins/providers/gcs/config.go

```diff
@@ -0,0 +1,194 @@
+	BucketRoleWriter        = "roles/storage.legacyBucketWriter"
+	BucketRoleObjectOwner   = "roles/storage.legacyObjectOwner"
+	BucketRoleObjectReader  = "roles/storage.legacyObjectReader"
+	BucketRoleObjectAdmin   = "roles/storage.objectAdmin"
+	BucketRoleObjectCreator = "roles/storage.objectCreator"
+	BucketRoleObjectViewer  = "roles/storage.objectViewer"
+
+	AccountTypeUser           = "user"
+	AccountTypeServiceAccount = "serviceAccount"
+	AccountTypeGroup          = "group"
+	AccountTypeDomain         = "domain"
+)
+
+type Config struct {
+	ProviderConfig *domain.ProviderConfig
+	valid          bool
+
+	crypto    domain.Crypto
+	validator *validator.Validate
+}
+
+func (c *Config) EncryptCredentials() error {
+	if err := c.parseAndValidate(); err != nil {
+		return err
+	}
+
+	credentials, ok := c.ProviderConfig.Credentials.(*Credentials)
+	if !ok {
+		return ErrInvalidCredentialsType
+	}
+
+	if err := credentials.Encrypt(c.crypto); err != nil {
+		return err
+	}
+
+	c.ProviderConfig.Credentials = credentials
+	return nil
+}
+
+type Credentials struct {
+	ServiceAccountKey string `json:"service_account_key" mapstructure:"service_account_key" validate:"required,base64"`
+	ResourceName      string `json:"resource_name" mapstructure:"resource_name" validate:"required"`
+}
+
+func (c *Credentials) Decrypt(decryptor domain.Decryptor) error {
+	if c == nil {
+		return ErrUnableToDecryptNilCredentials
+	}
+
+	decryptedServiceAccount, err := decryptor.Decrypt(c.ServiceAccountKey)
+	if err != nil {
+		return err
+	}
+
+	c.ServiceAccountKey = decryptedServiceAccount
+	return nil
+}
+
+func (c *Credentials) Encrypt(encryptor domain.Encryptor) error {
+	if c == nil {
+		return ErrUnableToEncryptNilCredentials
+	}
+
+	encryptedServiceAccount, err := encryptor.Encrypt(c.ServiceAccountKey)
+	if err != nil {
+		return err
+	}
+
+	c.ServiceAccountKey = encryptedServiceAccount
+	return nil
+}
+
+type Permission string
+
+func NewConfig(pc *domain.ProviderConfig, crypto domain.Crypto) *Config {
+	return &Config{
+		ProviderConfig: pc,
+		validator:      validator.New(),
+		crypto:         crypto,
+	}
+}
+
+func (c *Config) ParseAndValidate() error {
+	return c.parseAndValidate()
+}
+
+func (c *Config) parseAndValidate() error {
+	if c.valid {
+		return nil
+	}
+
+	validationError := []error{}
+
+	credentials, err := c.validateCredentials(c.ProviderConfig.Credentials)
+	if err != nil {
+		validationError = append(validationError, err)
+	} else {
+		c.ProviderConfig.Credentials = credentials
+	}
+
+	for _, r := range c.ProviderConfig.Resources {
+		if err := c.validateResourceConfig(r); err != nil {
+			validationError = append(validationError, err)
+		}
+	}
+
+	if len(validationError) > 0 {
+		errorStrings := []string{}
+		for _, err := range validationError {
+			errorStrings = append(errorStrings, err.Error())
+		}
+		return errors.New(strings.Join(errorStrings, "\n"))
+	}
+
+	c.valid = true
+	return nil
```

Comment:
this valid boolean is also not needed if this validation is only being done once when we read the config. we can make this function a pure function by not changing the state.

Reply:
we were using the same in other providers in guardian. Asked this from @rahmatrhd and he told this boolean value will be required if we're calling the parseAndValidate function apart from the create config function. So the entire computation could be avoided. Will look into it once again

Reply:
@Chief-Rishab need to check again if this provider plugin requires that. In bigquery provider it was needed

Reply:
@mabdh updated the changes in https://github.com/odpf/guardian/pull/227/commits/b70e3ec4aa8e55ff4938dd43d641c1ce007e8bc3

Reply:
@mabdh can we close this conversation?